### ISSUE TITLE
Support JSON data type in SQL Server (mssql)

### ISSUE THREAD

**Issue by lukeify** [sequelize/sequelize#8761]
JSON is an available data type in SQL server as of this year: https://docs.microsoft.com/en-us/sql/relational-databases/json/json-data-sql-server

As per Sequelize docs, SQL Server JSON support isn't available:

> Sequelize.JSON // JSON column. PostgreSQL, SQLite and MySQL only.

(Taken from http://docs.sequelizejs.com/manual/tutorial/models-definition.html#data-types)

**Comment by mccabemj** [comment]
As far as I understand it, there is no JSON data type for SQL Server. What was introduced in MSSQL 2016+ was native JSON support. The data type would stay as `NVARCHAR` but there are now built-in functions and operators such as `JSON_QUERY` which can return an object from a JSON string.

`SELECT JSON_VALUE(columnName, '$.personalDetails.Name') AS Name FROM Users`

I'm happy to take this one and start working on it @sushantdhiman

**Comment by sushantdhiman** [comment]
Go ahead @mccabemj

**Comment by mccabemj** [comment]
Actually, after looking into this, is already possible out of the box. To a point.

using .fn, we can do all the functions. For example

```
User.findAll({
  attributes: [[ sequelize.fn('JSON_QUERY', sequelize.col('userDetails'), '$.address'), 'users address']]
})
```

All we would need to do is document it, which I will. (It would be useful to know where you'd like this to go. Should it be in Querying/Attributes or its own section in Querying, or elsewhere?)

The issue here is that the returned result will be a string, not a JSON object. Thats fine for JSON_VALUE or ISJSON, not for JSON_QUERY.

The way to implement the returned result as JSON would be to use "FOR JSON" at the end of the query. Sadly, this gives the column a random name, not the name you set like `AS 'user address'` so using it at the moment makes the column ignored in model.js build as its not in the options.attributes.

We could simply add a new function like "convertToJson: true" to add the "FOR JSON PATH" to the end of the query, but can I have some suggestions as to what to do with the auto generated column name? model.js explicitly uses the attributes to return values, and this auto generated column name is obviously ignored.

**Comment by sushantdhiman** [comment]
MSSQL JSON support looks rather limited to me. Lets hold off until they actually support native JSON type. Meanwhile a section in querying probably our best option.

Please add a section JSON after Operators in "Querying" docs. Here a sub section MSSQL should explain special case for MSSQL and how to query for JSON data

**Comment by mhombach** [comment]
I'm somewhat new to Sequelize but isn't the goal of this ORM to make Model-Definitions that are valid regardless of the DB-engine? In my opinion even if MSSQL doesn't support JSON at all we should make it at least "valid" (even if it behaves not very nice in performance) and for example store the JSON as STRING and create some functions that query that json-object after getting the complete string from the DB. This way, even if this could be bad in performance for big JSON-objects, you could  still use the JSON-type for MSSQL/MySQL... Or am i missing something important here? Like i said, i'm new :)

**Comment by mccabemj** [comment]
@mhombach you are correct. However, in this instance, the work involved to 'fake' a JSON data type would be quite large. It is then masking the fact its not doing what is expected. 

For example, this entire issue was started by someone asking to include the JSON data type that Microsoft added, even though they didn't. Even I thought they had until I researched further. Many many people would start using Sequelize assuming they have a full JSON data type available to use with mssql when they don't really. 

However, the main reason I think that shouldn't be done, is related to getting JSON back out. The only way to get the db engine to do it is to add `FOR JSON`. This then takes every selected value and puts it into a single JSON object. Definitely not what is expected. So what we would end up doing is parsing the JSON within the library for the JSON columns. With all of that, what you are really developing is almost an entire fake of JSON support even though almost none of it is done by the DB engine itself. 

@sushantdhiman its up to you though. @mhombach has a valid point

**Comment by mccabemj** [comment]
Documentation updated in PR #8867

**Comment by sushantdhiman** [comment]
@mhombach As pointed out MSSQL just provides some helpers to work with stringified json stored in a column that is essentially just `NVARCHAR`. There is no native JSON type.

My real concern is choices / workarounds to support this. We have to provide same querying ability as other dialects. Other dialect offers JSON field where for MSSQL we need to handle special case for NVARCHAR. I would like to keep this simple for now.

In future if there is native JSON support from MSSQL we will surely add that to Sequelize.

**Comment by mhombach** [comment]
@sushantdhiman I wasn't talking about MSSQL although this issue was about it in the first place. I was talking about general support of JSON between all the engines which are supported in Sequelize.

And i fully understand your concerns about the choices, allthough i wonld not call it "workaround" if it is a feature that is implemented a clean way.

Also you are right that the JSON would be just a string/navchar within the MSSQL-DB, which needs to get processed by the library before returning it to the programmer/user. But what i love about the idea of it beeing implemented is that you truly can change your DB-engine midway and don't need to touch and change whole model-structures.

I completely share the opinion from @mccabemj :
> Many many people would start using Sequelize assuming they have a full JSON data type available to use with mssql when they don't really.

But you should still, in my opinion, leave it to the programmers if they want to use some sort of non-native JSON-support.
So... i would suggest that if one is using JSON-Models with the MSSQL-dialect/engine, that an error should be thrown in console, that there is no native JSON-support for MSSQL. BUT (!) if one want's to still enable this feature, he/she should place some option-variable like `enable_force_json_support = true` to skip this warning/error and let Sequelize use the new feature.
You never know what project will use this ORM and maybe they want to switch to MSSQL-engine to test some things and the only thing that is in the way is a small JSON-object with 2 little objects in some far far away archive-table. In this case the programmers should be able to run their project on MSSQL, even without native JSON-support.
Would this be something we can talk about?

**Comment by sushantdhiman** [comment]
@mhombach If a clean solution is presented which can support most case from our usual JSON API ([a](https://github.com/sequelize/sequelize/blob/master/test/integration/model/json.test.js), [b](https://github.com/sequelize/sequelize/blob/master/test/integration/model/json.test.js)) I will happily merge that PR :)

**Comment by papb** [comment]
@mhombach Would you still be interested in pursuing this?

**Comment by ShaharHD** [comment]
Hi,
joining a bit late to the party, but I would also suggest adding the following:

```typescript
     queryInterface.addConstraint(table, ['SomeJSONField'], {
        type: 'CHECK',
        where: Sequelize.where(Sequelize.fn('ISJSON', Sequelize.col('SomeJSONField')), { [Op.eq]: 1 }),
      });
```

This will insure the data you are reading and writing is actually JSON.

IMHO, as far as the underlying datatype - `NVARCHAR(MAX)` which is `Sequelize.TEXT` for that matter - as per Microsoft docs.

**Comment by mhombach** [comment]
@papb Sorry for the 1 year late response, missed your question somehow. I think i won't pursue this any further.
It's getting heavily worse on performance to do what i was thinking about some time ago.
The order of attributes doesn't matter to JSON-spec, so it gets extremely painfull to implement queries/searches on those fields.
I think if you are storing data that grows on JSON structures where you want to do queries on, one should consider an additional microservice using a NoSQL database. If the intend is just to store the JSON and retrieve it sometimes but never change or search on it, then this can just be stored as stringified JSON as plain text or as blob.
So from this point, i see no further reason why TypeORM should implement this :)
